### ISSUE TITLE
Failed to compile master on MacOS

### ISSUE THREAD

**Issue by tomgs** [FedericoPonzi/Horust#37]
#### Version Information:

* Version: 0.11.1

Macbook running Catalina 10.15.5, rustc 1.44.1 (c7087fe00 2020-06-17)
#### Description:
 

Compiling fails, giving:
```
 Compiling horust v0.1.1 (/Users/t0mgs/IdeaProjects/Horust)
error[E0432]: unresolved imports `libc::prctl`, `libc::PR_SET_CHILD_SUBREAPER`

src/horust/mod.rs:14:12
   |
14 | use libc::{prctl, PR_SET_CHILD_SUBREAPER};
   |            ^^^^^  ^^^^^^^^^^^^^^^^^^^^^^ no `PR_SET_CHILD_SUBREAPER` in the root
   |            |
   |            no `prctl` in the root

error[E0425]: cannot find function `execvpe` in module `nix::unistd`
   --> src/horust/runtime/process_spawner.rs:149:18
    |
149 |     nix::unistd::execvpe(program_name.as_ref(), arg_cptr.as_ref(), env_cptr.as_ref())?;
    |                  ^^^^^^^ help: a function with a similar name exists: `execve`
    | 
   ::: /Users/t0mgs/.cargo/registry/src/github.com-1ecc6299db9ec823/nix-0.16.1/src/unistd.rs:734:1
    |
734 | pub fn execve(path: &CStr, args: &[&CStr], env: &[&CStr]) -> Result<Void> {
    | ------------------------------------------------------------------------- similarly named function `execve` defined here

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0425, E0432.
For more information about an error, try `rustc --explain E0425`.
error: could not compile `horust`.
```

Haven't actually figured out what these deps are, just wanted to persist this if anyone else has the same issue.
@FedericoPonzi gettin' late round these parts, if you can take a look that would be great.

**Comment by FedericoPonzi** [comment]
Unfortunately as of today, horust will only compile on Linux.
Libc contains low bindings to *nix's libc, and the nix crate is a "safe" wrapper around those low level bindings. 
If for example you look here:
https://docs.rs/nix/0.17.0/src/nix/unistd.rs.html#772
```
#[cfg(any(target_os = "haiku",
          target_os = "linux",
          target_os = "openbsd"))]
````
This is missing the "macos" target.
I'm not sure if it would make sense to support Mac in the near future, but if you wish you can try to add it but I'm not sure how much effort would be required to implement it.

**Comment by tomgs** [comment]
Well, it feels to me that compiling a project you're working on # is probably a good thing to have! ;)

I could set up a small script to run the compilation on a container and run horust inside it, but that feels like a lot of "going around" the problem.

Let me dive a bit deeper to see whether there's already an attempt at making it work for Darwin and see where we go from there.

Haven't dove into the code yet, but is that crate vital or could we swap it?

**Comment by FedericoPonzi** [comment]
> Well, it feels to me that compiling a project you're working on # is probably a good thing to have! ;)

Indeed!

> I could set up a small script to run the compilation on a container and run horust inside it, but that feels like a lot of "going around" the problem.

I think a Makefile which wraps cargo would work here. "make build|test|run". There is already a dockerfile, so it should be something quick to do (modulo the ccache which is not there at this moment).

> Let me dive a bit deeper to see whether there's already an attempt at making it work for Darwin and see where we go from there.
> 
> Haven't dove into the code yet, but is that crate vital or could we swap it?

I don't think so. The supervision stuff could probably be abstracted away from the os, but some bits are kind of "low level" so they need to know some details of the os in order to work properly. Also, nix /libc are supported on macos, so I think it's just a matter of replacing linux specific function to either more generic functions (e.g. the execvpe error you got), or provide substitutions for Mac.
I think you could search for libc/nix used functions, and check the docs for each function and try to understand if they're compatible with Mac or how to integrate/replace them. In the best case scenario, as soon as you fix those two errors everything might build properly ([this](https://stackoverflow.com/questions/59922661/why-doesnt-unistd-h-execvpe-work-on-macos) exaplins the issue related with execvpe).

Also, if you want to go with this second option, you might want to update the CI to build also for macos, create a pr with all the changes. I could also help if we have ci and the pr ready.

**Comment by tomgs** [comment]
Alright, sounds reasonable! Will take a look tonight

**Comment by tomgs** [comment]
Looking at `libc`'s [crate docs](https://crates.io/crates/libc) it seems that there's a script that's used in CI, intended to make sure the currently installed rust toolchain can compile for target X. I'll get back to this in a second.

Let's first tackle the `libc` issue. I can see `libc` compiling properly, but after compilation `prctl` and `PR_SET_CHILD_SUBREAPER` are not available (note that the latter is an _optional_ first argument of the former).

Here's the usage of both in the current branch: https://github.com/FedericoPonzi/Horust/blob/e38dc967dd05676f8a19633a7e0371a08245eaa1/src/horust/mod.rs#L54-L58

Looking at the docs for the _crate's_ `prctl` [here](https://docs.rs/libc/0.2.72/libc/constant.PR_SET_CHILD_SUBREAPER.html) is not super useful, since it's just a binding to the low-level function. Looking at the _actual_ `prctl` docs [here](https://man7.org/linux/man-pages/man2/prctl.2.html) shows that it's used to control the calling process / thread. Specifically, when provided the `PR_SET_CHILD_SUBREAPER` (from the docs):

              If arg2 is nonzero, set the "child subreaper" attribute of the
              calling process; if arg2 is zero, unset the attribute.

              A subreaper fulfills the role of init(1) for its descendant
              processes.  When a process becomes orphaned (i.e., its
              immediate parent terminates), then that process will be
              reparented to the nearest still living ancestor subreaper.
              Subsequently, calls to getppid(2) in the orphaned process will
              now return the PID of the subreaper process, and when the
              orphan terminates, it is the subreaper process that will
              receive a SIGCHLD signal and will be able to wait(2) on the
              process to discover its termination status.

Awesome SE answer about subreapers [here](https://unix.stackexchange.com/questions/250153/what-is-a-subreaper-process), btw.

Since I _can_ see the `libc` compiling properly, yet I don't have any `prctl` in the root, I took a look at `libc` crate's [available platforms page](https://rust-lang.github.io/libc/#platform-specific-documentation), and it does not look like `darwin` or `macos` is there.

I need to go down the rabbit hole a bit more, but it's getting late again... time flies when you're having fun:)

(Also, I know this is all basic stuff - but these types of issues often turn out to be great blog post prompts, so I'm pre-doing all the work now:) ).

**Comment by tomgs** [comment]
Back at it now.

Even if I end up fixing the `execvpe` issue, I'm still left with `prctl` which doesn't exist on Mac OS, per [here](https://stackoverflow.com/questions/13617693/how-can-i-install-linux-sys-prctl-h-to-lion).

I think I'm gonna focus on getting this compiled in a container, and see how painful it is to develop with.

**Comment by FedericoPonzi** [comment]
I would like to provide few interesting points:
 * Creating an init system compatible with all the *nix systems is hard if not impossible. It would probably require many flags to work properly on every system, and I don't think it's something we should bother with at this time.
 * `prctl` is for for using some linux kernel specific features which are not present in Macos. It's also one of the main reasons why when you run docker on Mac, under the hood it spawns a linux VM on which it will run your containers.
 * In this particular instance, `prctl` is used for enabling the `PR_SET_CHILD_SUBREAPER`. I though it would have been useful for people who wanted to run horust as a normal process. But it's not really needed, everything should work as fine even if you put it behind a not(mac) flag.
 * Since you might encounter other issues and for now we're foremost targeting docker containers, I think it would be easier, faster and better if you could stick with the docker container. I would advise to update the makefile with your commands.

**Comment by tomgs** [comment]
![Oh my my](https://y.yarn.co/a22dce0a-b9ff-46ba-b54f-108539796878_text_hi.gif)

No Mac specifics for me, then. Back soon with a spruced-up Makefile.

**Comment by tomgs** [comment]
So, unsurprisingly, just building the image every time I make a change to a file takes forever (12+ minutes to be exact, between updating the crate index, downloading crates and compiling them). Need a way to cache compiled crates in between builds, and (possibly) refrain from updating the crates index as I go along.

https://stackoverflow.com/questions/58473606/cache-rust-dependencies-with-docker-build is a good start, and the follow up https://blog.mgattozzi.dev/caching-rust-docker-builds/ that I got from there. Hacking it in now.

**Comment by tomgs** [comment]
Before diving too deep here, check out https://benjamincongdon.me/blog/2019/12/04/Fast-Rust-Docker-Builds-with-cargo-vendor/

**Comment by tomgs** [comment]
Also - https://github.com/denzp/cargo-wharf for "magic" rust image dep cache builder

**Comment by tomgs** [comment]
(fixed by #41)
